### PR TITLE
SPU LLVM: Recognize ROTYQBYI and ROTQBI rotation pattern

### PR DESCRIPTION
**Pull request by Whatcookie**
ROTQBYBI shifts a vector by bytes, and ROTQBI shifts a vector by bits. We can recognize this pattern and shift by bytes and bits at the same time. Turns into just 1 instruction with AVX-512 (VPSHRDVD), but the AVX2 solution is also faster than before as well.

AVX-512
Before:
image

After:
image

AVX2
Before:
image

After:
image

### CODE REVIEW

**Comment by digant73** [comment]
do you have some stats about the improvements compared to current implementation?

**Comment by Whatcookie** [comment]
> do you have some stats about the improvements compared to current implementation?

It's not going to be a large improvement overall. I found the pattern while looking at hotspots in Dark Souls. Maybe some game benefits a lot.

File: rpcs3/Emu/Cell/SPULLVMRecompiler.cpp [hunk 1]
```diff
@@ -4813,6 +4822,38 @@ class spu_llvm_recompiler : public spu_recompiler_base, public cpu_translator
...
 		const auto a = get_vr(op.ra);
+		const auto ax = get_vr<u8[16]>(op.ra);
+		const auto bx = get_vr<u8[16]>(op.rb);
+
+		// Combined bit and bytes shift
+		if (auto [ok, v0, v1] = match_expr(ax, rotqbybi(match<u8[16]>(), match<u8[16]>())); ok && v1.eq(bx))
+		{
+			const auto b32 = get_vr<s32[4]>(op.rb);
+			// Is the rotate less than 31 bits?
+			if (auto k = get_known_bits(b32); !!(k.Zero & 0x60))
```
Comment (elad335) [inline rpcs3/Emu/Cell/SPULLVMRecompiler.cpp:4833]: It was checking for any of them being 0 which is wrong.
```suggestion
			if (auto k = get_known_bits(b32); (~k.Zero & 0x60) == 0)
```

**Comment by Augusto7743** [comment]
That PR is only for AVX2 and AVX-512 ?
Does problems for CPU having only AVX1 ?
Have an nice day.

**Comment by Hamof1** [comment]
If you have a CPU from 2011-2013, it's probably time for an upgrade.

**Pull request merged**